Gh-521: Document Gremlin integration in the REST API (#522)
gchq/gaffer-doc: docs/administration-guide/security/user-control.md

# User Authentication

!!! info "Work in Progress"

    This page is under construction.

The security layer for Gaffer is currently only enforced by the REST API. We
recommend restricting users such that they do not have access to the cluster and

> by only allowing users to connect to Gaffer via the REST API.

Reviewer: Not sure on this sentence phrasing, the bit after and could do with being clearer.

In the REST API the `User` object is constructed via a [`UserFactory`](https://gchq.github.io/Gaffer/uk/gov/gchq/gaffer/rest/factory/UserFactory.html).

> Under the Spring REST API this has an abstract implementation class

Reviewer: ```suggestion
In the Spring REST API an abstract implementation of this class is used,
```

> `AbstractUserFactory` to allow passing HTTP headers to be used for

Reviewer: ```suggestion
`AbstractUserFactory`, which is then used in the passing of HTTP headers for
```

Think these comments make it a bit clearer what you are trying to say here. Feel free to re-write however though

authentication.

Currently, we only provide one implementation of this, the `UnknownUserFactory`
which will always just return `new User()`. The `UnknownUserFactory` is the default
factory, to specify the user factory class define the

> `gaffer.user.factory.class` [REST property](../gaffer-config/config.md#application-properties).

Reviewer: This paragraph reads a little weird. Maybe just simplify to something like:
```
Currently, there is a single default implementation this; the `UnknownUserFactory` which simply returns a `new User()`. 
To specify the user factory class define the`gaffer.user.factory.class` [REST property](../gaffer-config/config.md#application-properties).
```

## Writing a User Factory

> To authenticate your users, you will need to extend the `AbstractUserFactory` class

Reviewer: ```suggestion
To authenticate your users you will need to extend the `AbstractUserFactory` class
```

# User Authentication

!!! info "Work in Progress"

    This page is under construction.

The user authentication layer for Gaffer is currently only enforced by the REST
API. We recommend restricting users such that they do not have access to the
underlying Java API so that all queries are authenticated and executed via the
REST API.

In the REST API the `User` object is constructed via a [`UserFactory`](https://gchq.github.io/Gaffer/uk/gov/gchq/gaffer/rest/factory/UserFactory.html).
In the Spring REST API an abstract implementation of this class is used,
`AbstractUserFactory`, which is then used in the passing of HTTP headers for
authentication.

Currently, there is a single default implementation of this; the
`UnknownUserFactory` which simply returns a new `User` with `UNKNOWN` as the

> user ID. The `UnknownUserFactory` is the default factory, to specify the user

Reviewer: ```suggestion
user ID. To specify the user
```
Not sure you need to repeat that UnknownUserFactory is the default as you have already said that in the previous sentence
